From 87981327563bd9cd2929930afea904b92d6252f1 Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Thu, 25 Sep 2025 17:43:28 +0800
Subject: [PATCH 128/128] mtk: mac80211: support receiving link_id from
 broadcast twt nl80211 msg

In MLD mode, hostapd may send a BTWT nl80211 message with a link ID.
mac80211 will include the link ID in struct cfg80211_broadcast_settings,
informing the underlying driver of the link ID for this BTWT schedule.

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
---
 include/net/cfg80211.h       |  6 ++++--
 include/net/mac80211.h       |  5 +++--
 include/uapi/linux/nl80211.h |  2 ++
 net/mac80211/cfg.c           |  8 ++++----
 net/mac80211/driver-ops.h    | 13 +++++++------
 net/mac80211/s1g.c           |  4 ++--
 net/mac80211/trace.h         | 21 +++++++++++++--------
 net/wireless/nl80211.c       | 14 +++++++++++---
 net/wireless/rdev-ops.h      | 13 ++++++-------
 net/wireless/trace.h         | 20 ++++++++++++--------
 10 files changed, 64 insertions(+), 42 deletions(-)

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index 8e8a26fb..44c0dca5 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -318,6 +318,7 @@ struct ieee80211_rate {
  *	to this broadcast TWT Parameter set are present
  * @dl_tid_bmp: latency sensitive traffic streams in the DL direction.
  * @ul_tid_bmp: latency sensitive traffic streams in the UL direction.
+ * @link_id: the MLO link id or -1 for non-MLD
  */
 struct cfg80211_broadcast_twt_settings {
 	u8 id;
@@ -5276,9 +5277,10 @@ struct cfg80211_ops {
 	int	(*set_epcs)(struct wiphy *wiphy, struct net_device *dev,
 			    bool val);
 	void	(*skip_cac)(struct wireless_dev *wdev, unsigned int link_id);
-	int	(*add_btwt)(struct wiphy *wiphy, struct net_device *dev,
+	int	(*add_btwt)(struct wiphy *wiphy, struct net_device *dev, u8 link_id,
 			    struct cfg80211_broadcast_twt_settings *settings);
-	int	(*del_btwt)(struct wiphy *wiphy, struct net_device *dev, u8 id);
+	int	(*del_btwt)(struct wiphy *wiphy, struct net_device *dev,
+			    u8 link_id, u8 id);
 };
 
 /*
diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 5d4b5dbf..a96b988c 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -4933,8 +4933,9 @@ struct ieee80211_ops {
 				    struct ieee80211_sta *sta,
 				    struct ieee80211_twt_teardown *td);
 	int (*add_btwt)(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
-			struct cfg80211_broadcast_twt_settings *btwt);
-	int (*del_btwt)(struct ieee80211_hw *hw, struct ieee80211_vif *vif, u8 id);
+			u8 link_id, struct cfg80211_broadcast_twt_settings *btwt);
+	int (*del_btwt)(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
+			u8 link_id, u8 id);
 	int (*set_radar_background)(struct ieee80211_hw *hw,
 				    struct cfg80211_chan_def *chandef);
 	int (*net_fill_forward_path)(struct ieee80211_hw *hw,
diff --git a/include/uapi/linux/nl80211.h b/include/uapi/linux/nl80211.h
index 6e3fd1a2..029a44ff 100644
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -8563,6 +8563,7 @@ enum nl80211_crit_update_event {
  * 	TWT scheduling AP.
  * @NL80211_BROADCAST_TWT_ATTR_PERSISTENCE: Number of TBTT before this
  * 	BTWT is end. Set 255 for a persist BTWT that never ends.
+ * @NL80211_BROADCAST_TWT_ATTR_LINK_ID: MLO link ID.
  * @__NL80211_BROADCAST_TWT_ATTR_LAST: Internal
  * @NL80211_BROADCAST_TWT_ATTR_MAX: Highest attribute
  */
@@ -8580,6 +8581,7 @@ enum nl80211_broadcast_twt_attributes {
 	NL80211_BROADCAST_TWT_ATTR_PERSISTENCE,
 	NL80211_BROADCAST_TWT_ATTR_DL_TID_BMP,
 	NL80211_BROADCAST_TWT_ATTR_UL_TID_BMP,
+	NL80211_BROADCAST_TWT_ATTR_LINK_ID,
 
 	/* keep last */
 	__NL80211_BROADCAST_TWT_ATTR_LAST,
diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 4e1b654e..075a4d1d 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -5842,7 +5842,7 @@ void ieee80211_crit_update_notify(struct ieee80211_vif *vif, unsigned int link_i
 EXPORT_SYMBOL_GPL(ieee80211_crit_update_notify);
 
 static int
-ieee80211_add_btwt(struct wiphy *wiphy, struct net_device *dev,
+ieee80211_add_btwt(struct wiphy *wiphy, struct net_device *dev, u8 link_id,
 		   struct cfg80211_broadcast_twt_settings *settings)
 {
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
@@ -5851,7 +5851,7 @@ ieee80211_add_btwt(struct wiphy *wiphy, struct net_device *dev,
 
 	lockdep_assert_wiphy(sdata->local->hw.wiphy);
 
-	ret = drv_add_btwt(local, sdata, settings);
+	ret = drv_add_btwt(local, sdata, link_id, settings);
 	if (!ret)
 		sdata->deflink.btwt_active_bitmap |= BIT(settings->id);
 
@@ -5859,7 +5859,7 @@ ieee80211_add_btwt(struct wiphy *wiphy, struct net_device *dev,
 }
 
 static int
-ieee80211_del_btwt(struct wiphy *wiphy, struct net_device *dev, u8 id)
+ieee80211_del_btwt(struct wiphy *wiphy, struct net_device *dev, u8 link_id, u8 id)
 {
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
 	struct ieee80211_local *local = sdata->local;
@@ -5867,7 +5867,7 @@ ieee80211_del_btwt(struct wiphy *wiphy, struct net_device *dev, u8 id)
 
 	lockdep_assert_wiphy(sdata->local->hw.wiphy);
 
-	ret =  drv_del_btwt(local, sdata, id);
+	ret = drv_del_btwt(local, sdata, link_id, id);
 	if (!ret)
 		sdata->deflink.btwt_active_bitmap &= ~BIT(id);
 
diff --git a/net/mac80211/driver-ops.h b/net/mac80211/driver-ops.h
index d3d31a90..79f6ec29 100644
--- a/net/mac80211/driver-ops.h
+++ b/net/mac80211/driver-ops.h
@@ -1680,7 +1680,7 @@ static inline int drv_twt_teardown_request(struct ieee80211_local *local,
 }
 
 static inline int drv_add_btwt(struct ieee80211_local *local,
-			       struct ieee80211_sub_if_data *sdata,
+			       struct ieee80211_sub_if_data *sdata, u8 link_id,
 			       struct cfg80211_broadcast_twt_settings *settings)
 {
 	int ret = -EOPNOTSUPP;
@@ -1690,16 +1690,17 @@ static inline int drv_add_btwt(struct ieee80211_local *local,
 	if (!check_sdata_in_driver(sdata))
 		return -EIO;
 
-	trace_drv_add_btwt(local, sdata, settings);
+	trace_drv_add_btwt(local, sdata, link_id, settings);
 	if (local->ops->add_btwt)
-		ret = local->ops->add_btwt(&local->hw, &sdata->vif, settings);
+		ret = local->ops->add_btwt(&local->hw, &sdata->vif, link_id, settings);
 	trace_drv_return_int(local, ret);
 
 	return ret;
 }
 
 static inline int drv_del_btwt(struct ieee80211_local *local,
-			       struct ieee80211_sub_if_data *sdata, u8 id)
+			       struct ieee80211_sub_if_data *sdata,
+			       u8 link_id, u8 id)
 {
 	int ret = -EOPNOTSUPP;
 
@@ -1708,9 +1709,9 @@ static inline int drv_del_btwt(struct ieee80211_local *local,
 	if (!check_sdata_in_driver(sdata))
 		return -EIO;
 
-	trace_drv_del_btwt(local, sdata, id);
+	trace_drv_del_btwt(local, sdata, link_id, id);
 	if (local->ops->del_btwt)
-		ret = local->ops->del_btwt(&local->hw, &sdata->vif, id);
+		ret = local->ops->del_btwt(&local->hw, &sdata->vif, link_id, id);
 	trace_drv_return_int(local, ret);
 
 	return ret;
diff --git a/net/mac80211/s1g.c b/net/mac80211/s1g.c
index ab87d19a..d5577b5d 100644
--- a/net/mac80211/s1g.c
+++ b/net/mac80211/s1g.c
@@ -143,8 +143,8 @@ ieee80211_s1g_handle_twt_demand_setup(struct ieee80211_sub_if_data *sdata,
 		return -ENOSPC;
 
 	settings.id = ffz(sdata->deflink.btwt_active_bitmap | BIT(0));
-
-	ret = drv_add_btwt(sdata->local, sdata, &settings);
+	/* FIXME: The link id shall be decided based on the effective link. */
+	ret = drv_add_btwt(sdata->local, sdata, 0, &settings);
 	if (ret)
 		return ret;
 
diff --git a/net/mac80211/trace.h b/net/mac80211/trace.h
index 850dd621..28d5babb 100644
--- a/net/mac80211/trace.h
+++ b/net/mac80211/trace.h
@@ -2559,10 +2559,10 @@ TRACE_EVENT(drv_twt_teardown_request,
 
 TRACE_EVENT(drv_add_btwt,
 	TP_PROTO(struct ieee80211_local *local,
-		 struct ieee80211_sub_if_data *sdata,
+		 struct ieee80211_sub_if_data *sdata, u8 link_id,
 		 struct cfg80211_broadcast_twt_settings *settings),
 
-	TP_ARGS(local, sdata, settings),
+	TP_ARGS(local, sdata, link_id, settings),
 
 	TP_STRUCT__entry(
 		LOCAL_ENTRY
@@ -2576,6 +2576,7 @@ TRACE_EVENT(drv_add_btwt,
 		__field(bool, flow_type)
 		__field(u8, recommendation)
 		__field(u8, persist)
+		__field(u8, link_id)
 	),
 
 	TP_fast_assign(
@@ -2590,40 +2591,44 @@ TRACE_EVENT(drv_add_btwt,
 		__entry->flow_type = settings->flow_type;
 		__entry->recommendation = settings->recommendation;
 		__entry->persist = settings->persist;
+		__entry->link_id = link_id;
 	),
 
 	TP_printk(
 		LOCAL_PR_FMT VIF_PR_FMT " id=%u min_wake_dur=%u "
 		"wake_intv_mantissa=%u wake_intv_exp=%u setup_cmd=%u trigger=%u"
-		"flow_type=%u recommendation=%u persist=%u",
+		"flow_type=%u recommendation=%u persist=%u link_id=%u",
 		LOCAL_PR_ARG, VIF_PR_ARG, __entry->id, __entry->min_wake_dur,
 		__entry->wake_intv_mantissa, __entry->wake_intv_exp,
 		__entry->setup_cmd, __entry->trigger, __entry->flow_type,
-		__entry->recommendation, __entry->persist
+		__entry->recommendation, __entry->persist, __entry->link_id
 	)
 );
 
 TRACE_EVENT(drv_del_btwt,
 	TP_PROTO(struct ieee80211_local *local,
-		 struct ieee80211_sub_if_data *sdata, u8 id),
+		 struct ieee80211_sub_if_data *sdata,
+		 u8 link_id, u8 id),
 
-	TP_ARGS(local, sdata, id),
+	TP_ARGS(local, sdata, link_id, id),
 
 	TP_STRUCT__entry(
 		LOCAL_ENTRY
 		VIF_ENTRY
 		__field(u8, id)
+		__field(u8, link_id)
 	),
 
 	TP_fast_assign(
 		LOCAL_ASSIGN;
 		VIF_ASSIGN;
 		__entry->id = id;
+		__entry->link_id = link_id;
 	),
 
 	TP_printk(
-		LOCAL_PR_FMT VIF_PR_FMT " id=%u",
-		LOCAL_PR_ARG, VIF_PR_ARG, __entry->id
+		LOCAL_PR_FMT VIF_PR_FMT " id=%u link_id=%u",
+		LOCAL_PR_ARG, VIF_PR_ARG, __entry->id, __entry->link_id
 	)
 );
 
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index ec304506..b19da1e6 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -578,6 +578,7 @@ nl80211_broadcast_twt_policy[NL80211_BROADCAST_TWT_ATTR_MAX + 1] = {
 	[NL80211_BROADCAST_TWT_ATTR_PERSISTENCE] = { .type = NLA_U8 },
 	[NL80211_BROADCAST_TWT_ATTR_DL_TID_BMP] = { .type = NLA_U8 },
 	[NL80211_BROADCAST_TWT_ATTR_UL_TID_BMP] = { .type = NLA_U8 },
+	[NL80211_BROADCAST_TWT_ATTR_LINK_ID] = { .type = NLA_U8 },
 };
 
 static const struct nla_policy nl80211_policy[NUM_NL80211_ATTR] = {
@@ -18065,6 +18066,7 @@ nl80211_add_broadcast_twt(struct sk_buff *skb, struct genl_info *info)
 	struct wireless_dev *wdev = dev->ieee80211_ptr;
 	struct nlattr *tb[NL80211_BROADCAST_TWT_ATTR_MAX + 1];
 	struct cfg80211_broadcast_twt_settings settings;
+	u8 link_id = 0;
 	int ret;
 
 	if (wdev->iftype != NL80211_IFTYPE_AP)
@@ -18108,7 +18110,10 @@ nl80211_add_broadcast_twt(struct sk_buff *skb, struct genl_info *info)
 	if (tb[NL80211_BROADCAST_TWT_ATTR_UL_TID_BMP])
 		settings.ul_tid_bmp = nla_get_u8(tb[NL80211_BROADCAST_TWT_ATTR_UL_TID_BMP]);
 
-	return rdev_add_btwt(rdev, dev, &settings);
+	if (tb[NL80211_BROADCAST_TWT_ATTR_LINK_ID])
+		link_id = nla_get_u8(tb[NL80211_BROADCAST_TWT_ATTR_LINK_ID]);
+
+	return rdev_add_btwt(rdev, dev, link_id, &settings);
 }
 
 static int
@@ -18118,7 +18123,7 @@ nl80211_del_broadcast_twt(struct sk_buff *skb, struct genl_info *info)
 	struct net_device *dev = info->user_ptr[1];
 	struct wireless_dev *wdev = dev->ieee80211_ptr;
 	struct nlattr *tb[NL80211_BROADCAST_TWT_ATTR_MAX + 1];
-	u8 id;
+	u8 id, link_id = 0;
 	int ret;
 
 	if (wdev->iftype != NL80211_IFTYPE_AP)
@@ -18135,7 +18140,10 @@ nl80211_del_broadcast_twt(struct sk_buff *skb, struct genl_info *info)
 
 	id = nla_get_u8(tb[NL80211_BROADCAST_TWT_ATTR_ID]);
 
-	return rdev_del_btwt(rdev, dev, id);
+	if (tb[NL80211_BROADCAST_TWT_ATTR_LINK_ID])
+		link_id = nla_get_u8(tb[NL80211_BROADCAST_TWT_ATTR_LINK_ID]);
+
+	return rdev_del_btwt(rdev, dev, link_id, id);
 }
 
 #define NL80211_FLAG_NEED_WIPHY		0x01
diff --git a/net/wireless/rdev-ops.h b/net/wireless/rdev-ops.h
index 9dc61a1f..5a11dc50 100644
--- a/net/wireless/rdev-ops.h
+++ b/net/wireless/rdev-ops.h
@@ -1651,15 +1651,15 @@ rdev_skip_cac(struct cfg80211_registered_device *rdev,
 
 static inline int
 rdev_add_btwt(struct cfg80211_registered_device *rdev,
-	      struct net_device *dev,
+	      struct net_device *dev, u8 link_id,
 	      struct cfg80211_broadcast_twt_settings *settings)
 {
 	struct wiphy *wiphy = &rdev->wiphy;
 	int ret = -EOPNOTSUPP;
 
-	trace_rdev_add_btwt(wiphy, dev, settings);
+	trace_rdev_add_btwt(wiphy, dev, link_id, settings);
 	if (rdev->ops->add_btwt)
-		ret = rdev->ops->add_btwt(wiphy, dev, settings);
+		ret = rdev->ops->add_btwt(wiphy, dev, link_id, settings);
 	trace_rdev_return_int(wiphy, ret);
 
 	return ret;
@@ -1667,15 +1667,14 @@ rdev_add_btwt(struct cfg80211_registered_device *rdev,
 
 static inline int
 rdev_del_btwt(struct cfg80211_registered_device *rdev,
-	      struct net_device *dev,
-	      u8 id)
+	      struct net_device *dev, u8 link_id, u8 id)
 {
 	struct wiphy *wiphy = &rdev->wiphy;
 	int ret = -EOPNOTSUPP;
 
-	trace_rdev_del_btwt(wiphy, dev, id);
+	trace_rdev_del_btwt(wiphy, dev, link_id, id);
 	if (rdev->ops->del_btwt)
-		ret = rdev->ops->del_btwt(wiphy, dev, id);
+		ret = rdev->ops->del_btwt(wiphy, dev, link_id, id);
 	trace_rdev_return_int(wiphy, ret);
 
 	return ret;
diff --git a/net/wireless/trace.h b/net/wireless/trace.h
index b0fda4f5..35d7a047 100644
--- a/net/wireless/trace.h
+++ b/net/wireless/trace.h
@@ -4332,9 +4332,9 @@ TRACE_EVENT(rdev_skip_cac,
 
 TRACE_EVENT(rdev_add_btwt,
 	TP_PROTO(struct wiphy *wiphy,
-		 struct net_device *netdev,
+		 struct net_device *netdev, u8 link_id,
 		 struct cfg80211_broadcast_twt_settings *settings),
-	TP_ARGS(wiphy, netdev, settings),
+	TP_ARGS(wiphy, netdev, link_id, settings),
 	TP_STRUCT__entry(
 		WIPHY_ENTRY
 		NETDEV_ENTRY
@@ -4347,6 +4347,7 @@ TRACE_EVENT(rdev_add_btwt,
 		__field(bool, flow_type)
 		__field(u8, recommendation)
 		__field(u8, persist)
+		__field(u8, link_id)
 	),
 	TP_fast_assign(
 		WIPHY_ASSIGN;
@@ -4360,33 +4361,36 @@ TRACE_EVENT(rdev_add_btwt,
 		__entry->flow_type = settings->flow_type;
 		__entry->recommendation = settings->recommendation;
 		__entry->persist = settings->persist;
+		__entry->link_id = link_id;
 	),
 	TP_printk(WIPHY_PR_FMT NETDEV_PR_FMT " id=%u min_wake_dur=%u "
 		"wake_intv_mantissa=%u wake_intv_exp=%u setup_cmd=%u trigger=%u"
-		"flow_type=%u recommendation=%u persist=%u",
+		"flow_type=%u recommendation=%u persist=%u link_id=%u",
 		  WIPHY_PR_ARG, NETDEV_PR_ARG, __entry->id, __entry->min_wake_dur,
 		__entry->wake_intv_mantissa, __entry->wake_intv_exp,
 		__entry->setup_cmd, __entry->trigger, __entry->flow_type,
-		__entry->recommendation, __entry->persist)
+		__entry->recommendation, __entry->persist, __entry->link_id)
 );
 
 TRACE_EVENT(rdev_del_btwt,
 	TP_PROTO(struct wiphy *wiphy,
 		 struct net_device *netdev,
-		 u8 id),
-	TP_ARGS(wiphy, netdev, id),
+		 u8 link_id, u8 id),
+	TP_ARGS(wiphy, netdev, link_id, id),
 	TP_STRUCT__entry(
 		WIPHY_ENTRY
 		NETDEV_ENTRY
 		__field(u8, id)
+		__field(u8, link_id)
 	),
 	TP_fast_assign(
 		WIPHY_ASSIGN;
 		NETDEV_ASSIGN;
 		__entry->id = id;
+		__entry->link_id = link_id;
 	),
-	TP_printk(WIPHY_PR_FMT NETDEV_PR_FMT " id=%u",
-		  WIPHY_PR_ARG, NETDEV_PR_ARG, __entry->id)
+	TP_printk(WIPHY_PR_FMT NETDEV_PR_FMT " id=%u link_id=%u",
+		  WIPHY_PR_ARG, NETDEV_PR_ARG, __entry->id, __entry->link_id)
 );
 #endif /* !__RDEV_OPS_TRACE || TRACE_HEADER_MULTI_READ */
 
-- 
2.45.2

